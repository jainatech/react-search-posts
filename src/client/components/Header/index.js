import React from 'react'
import PropTypes from 'prop-types'
import { Navbar, Form, FormControl } from 'react-bootstrap';

function Header(props) {
    const searchRef = React.createRef();

    const doSearch = () => {
        if (searchRef && searchRef.current && searchRef.current.value) {
            props.filterData(searchRef.current.value);
        }
    }

    return (
        <>
            <Navbar expand="lg" variant="light" bg="light" className="bg-light justify-content-between">
                <Navbar.Brand>{props.text}</Navbar.Brand>
                {props.search ?
                    <Form inline>
                        <FormControl ref={searchRef} type="text" placeholder="Search by title" className=" mr-sm-2" onChange={doSearch} />
                    </Form> : null}

            </Navbar>
        </>
    )
}

Header.propTypes = {
    text: PropTypes.string,
    filterData: PropTypes.func
}

export default Header

