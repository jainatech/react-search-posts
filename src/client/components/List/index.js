import React from 'react'
import PropTypes from 'prop-types'
import { Table } from 'react-bootstrap';
import ListItem from "../ListItem";

const List = ({ data, type }) => {
    switch (type) {
        case "home_posts_list":
            return (
                <Table>
                    <tbody>
                        {data && data.length ?
                            data.map((post, index) => {
                                return <ListItem key={`post-${post.id}`} item={post} index={index} type="home_posts_list" />;
                            }) : ''
                        }
                    </tbody>
                </Table>
            );
        default:
            return (<p>List</p>)
    }

}

List.propTypes = {
    data: PropTypes.array.isRequired,
    type: PropTypes.string
}

export default List

