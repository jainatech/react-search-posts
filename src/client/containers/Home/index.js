import React, { useState, useEffect } from 'react'
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Container, Row, Col, Pagination } from 'react-bootstrap';
import List from '../../components/List';
import { getPost } from '../../store/actions/posts';
import Header from '../../components/Header';


const Home = ({ posts, loading, getPost }) => {

    const [postsData, setPostsData] = useState(undefined);
    const [active, setActive] = useState(1);
    const limit = 10;

    useEffect(() => {
        console.log("POST",posts)
        if (posts && posts.length === 0) {
            getPost();
        }
    });


    const filterData = (searchValue) => {
        let result = [];
        if (searchValue) {
            setActive(1);
            result = posts.filter(post =>
                post.title.toLowerCase().indexOf(searchValue.toLowerCase()) > -1
            )
            setPostsData(result);
        } else {
            result = posts;
            setPostsData(result);
        }
        return result;
    };

    const getRenderList = () => {
        if (loading) {
            return <h4 className="text-center">Loading...</h4>;
        } else {
            let render = []
            if (postsData) {
                render = postsData;
            } else {
                render = posts;
            }
            const indexOfLast = active * limit;
            const indexOfFirst = indexOfLast - limit;
            console.log(render);
            render = render && render.length > 0 ? render.slice(indexOfFirst, indexOfLast) : [];
            return <List data={render} type="home_posts_list" />
        }

    }


    return (

        <>
            <Header text="Posts Search App" filterData={filterData} search />
            <Container fluid="lg">
                <Row>
                    <Col>
                        <Container>
                            <Row>
                                <Col>
                                    {getRenderList()}
                                </Col>
                            </Row>
                            <Row>
                                <Pagination>{Array(Math.ceil(postsData ? (postsData.length / limit) : (posts.length / limit))).fill().map((x, i) => i + 1).map(item => {
                                    return <Pagination.Item key={item} active={item === active} onClick={() => setActive(item)} >
                                        {item}
                                    </Pagination.Item>
                                })}</Pagination>
                            </Row>
                        </Container>
                    </Col>
                </Row>
            </Container>
        </>
    )
}

Home.propTypes = {
    posts: PropTypes.array,
    loading: PropTypes.bool,
    getPost: PropTypes.func
}

const mapStateToProps = (state) => ({
    posts: state.postsReducer.posts_data,
    loading: state.postsReducer.loading
})

const mapDispatchToProps = dispatch => bindActionCreators({
    getPost: getPost
}, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Home)


