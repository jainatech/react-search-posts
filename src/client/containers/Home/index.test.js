import React from 'react';
import { shallow, mount } from 'enzyme';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import Header from '../../components/Header';
import List from '../../components/List';
import Home from './index';
const mockStore = configureStore();
const userStore = mockStore({
  postsReducer: {
    posts_data: [{ "userId": 1, "id": 1, "title": "title1", "body": "body1" }, { "userId": 1, "id": 2, "title": "title2", "body": "body2" }],
    loading: true
  }
});

const createTestProps = (props) => ({
  ...props
})

describe('HOME module case--->', () => {
  it('Home match snapshot', () => {
    let props = createTestProps({
      posts: [{ "userId": 1, "id": 1, "title": "title1", "body": "body1" }, { "userId": 1, "id": 2, "title": "title2", "body": "body2" }],
      loading: true
    })
    let wrapper = mount(
      <Provider store={userStore}><Home {...props}></Home></Provider>);
    expect(wrapper).toMatchSnapshot();
    wrapper.unmount();
  });

  it('Home Renders without crashing', () => {
    const component = shallow(<Provider store={userStore}><Home /></Provider>);
    expect(component.find(Home)).toBeDefined();
  });

  it('Home run without crashing', () => {
    const userStore = mockStore({});
    const component = shallow(<Provider store={userStore}><Header /></Provider>, { store: userStore });
    expect(component.find(Header)).toBeDefined();
  });
});