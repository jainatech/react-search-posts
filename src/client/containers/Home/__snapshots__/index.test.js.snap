// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HOME module case---> Home match snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Home)
    loading={true}
    posts={
      Array [
        Object {
          "body": "body1",
          "id": 1,
          "title": "title1",
          "userId": 1,
        },
        Object {
          "body": "body2",
          "id": 2,
          "title": "title2",
          "userId": 1,
        },
      ]
    }
  >
    <Home
      getPost={[Function]}
      loading={true}
      posts={
        Array [
          Object {
            "body": "body1",
            "id": 1,
            "title": "title1",
            "userId": 1,
          },
          Object {
            "body": "body2",
            "id": 2,
            "title": "title2",
            "userId": 1,
          },
        ]
      }
    >
      <Header
        filterData={[Function]}
        search={true}
        text="Posts Search App"
      >
        <Navbar
          bg="light"
          className="bg-light justify-content-between"
          collapseOnSelect={false}
          expand="lg"
          variant="light"
        >
          <nav
            className="bg-light justify-content-between navbar navbar-expand-lg navbar-light bg-light"
          >
            <NavbarBrand>
              <span
                className="navbar-brand"
              >
                Posts Search App
              </span>
            </NavbarBrand>
            <Form
              inline={true}
            >
              <form
                className="form-inline"
              >
                <FormControl
                  className=" mr-sm-2"
                  onChange={[Function]}
                  placeholder="Search by title "
                  type="text"
                >
                  <input
                    className=" mr-sm-2 form-control"
                    onChange={[Function]}
                    placeholder="Search by title "
                    type="text"
                  />
                </FormControl>
              </form>
            </Form>
          </nav>
        </Navbar>
      </Header>
      <Container
        fluid="lg"
      >
        <div
          className="container-lg"
        >
          <Row
            noGutters={false}
          >
            <div
              className="row"
            >
              <Col>
                <div
                  className="col"
                >
                  <Container
                    fluid={false}
                  >
                    <div
                      className="container"
                    >
                      <Row
                        noGutters={false}
                      >
                        <div
                          className="row"
                        >
                          <Col>
                            <div
                              className="col"
                            >
                              <h4
                                className="text-center"
                              >
                                Loading...
                              </h4>
                            </div>
                          </Col>
                        </div>
                      </Row>
                      <Row
                        noGutters={false}
                      >
                        <div
                          className="row"
                        >
                          <ForwardRef>
                            <ul
                              className="pagination"
                            >
                              <PageItem
                                active={true}
                                activeLabel="(current)"
                                disabled={false}
                                key="1"
                                onClick={[Function]}
                              >
                                <li
                                  className="page-item active"
                                >
                                  <span
                                    className="page-link"
                                    disabled={false}
                                    onClick={[Function]}
                                  >
                                    1
                                    <span
                                      className="sr-only"
                                    >
                                      (current)
                                    </span>
                                  </span>
                                </li>
                              </PageItem>
                            </ul>
                          </ForwardRef>
                        </div>
                      </Row>
                    </div>
                  </Container>
                </div>
              </Col>
            </div>
          </Row>
        </div>
      </Container>
    </Home>
  </Connect(Home)>
</Provider>
`;
